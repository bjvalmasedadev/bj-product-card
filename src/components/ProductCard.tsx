import React, { CSSProperties, createContext } from 'react';
import useProducts from '../hooks/useProducts';
import {
  InitialValue,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  // children?: ReactElement | ReactElement[];
  className?: string;
  initialValue?: InitialValue;
  product: Product;
  style?: CSSProperties;
  value?: number;
  children: (args: ProductCardHandlers) => JSX.Element;
  onChange?: (args: OnChangeArgs) => void;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValue,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReach,
    reset,
  } = useProducts({ onChange, product, value, initialValue });

  return (
    <Provider
      value={{
        counter,
        maxCount,
        product,
        increaseBy,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReach,
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
