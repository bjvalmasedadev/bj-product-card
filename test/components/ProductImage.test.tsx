import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/product';

describe('Product Image', () => {
  test('Should render the component with the custom image', () => {
    const wrapper = renderer.create(<ProductImage img="http://img.png" />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Should render with the product image', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
