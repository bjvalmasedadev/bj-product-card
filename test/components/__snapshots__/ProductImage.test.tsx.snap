// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Image Should render the component with the custom image 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="http://img.png"
/>
`;

exports[`Product Image Should render with the product image 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="/coffee-mug.png"
  />
</div>
`;
